Tests/InstrumentationTests/TracingInstrumentTests.swift

```diff
@@ -114,12 +113,52 @@ struct OTSpan: Span {
 
     private(set) var links = [SpanLink]()
 
-    var attributes: SpanAttributes = [:] {
+    private var attributes: SpanAttributes = [:] {
         didSet {
             self.isRecording = !self.attributes.isEmpty
         }
     }
 
+    mutating func setAttribute(_ value: String, forKey key: String) {
+        self.attributes[key] = .string(value)
+    }
+
+    mutating func setAttribute(_ value: [String], forKey key: String) {
+        self.attributes[key] = .array(value.map(SpanAttribute.string))
+    }
+
+    mutating func setAttribute(_ value: Int, forKey key: String) {
+        self.attributes[key] = .int(value)
+    }
+
+    mutating func setAttribute(_ value: [Int], forKey key: String) {
+        self.attributes[key] = .array(value.map(SpanAttribute.int))
+    }
+
+    mutating func setAttribute(_ value: Double, forKey key: String) {
+        self.attributes[key] = .double(value)
+    }
+
+    mutating func setAttribute(_ value: [Double], forKey key: String) {
+        self.attributes[key] = .array(value.map(SpanAttribute.double))
+    }
+
+    mutating func setAttribute(_ value: Bool, forKey key: String) {
+        self.attributes[key] = .bool(value)
+    }
+
+    mutating func setAttribute(_ value: [Bool], forKey key: String) {
+        self.attributes[key] = .array(value.map(SpanAttribute.bool))
+    }
+
+    mutating func setAttribute(_ value: CustomStringConvertible, forKey key: String) {
+        self.attributes[key] = .stringConvertible(value)
+    }
```

Comment:
its more of a convenience method and does not define new type as such,

personally I'd add it to protocol extension (code for ilustration purposes, did not test it):

```swift
extension Span {
    public func setAnnotation(_ value: CustomStringConvertible, forKey key: String) {
       setAnnotation(String(describing: value), forKey: key)
    }
}
```